lib/script.js

```diff
@@ -40,106 +37,111 @@ Script.prototype.run = function(methodName, globals) {
-      client.write(EJSON.stringify({
-        source: _this.source,// the untrusted JS.
-        sourceAPI: _this.sourceAPI,// the trusted API.
+    const client = net.connect(this.sandcastle.getSocket(), () => {
+      let pack = EJSON.stringify({
+        source: this.source, // the untrusted JS.
+        sourceAPI: this.sourceAPI, // the trusted API.
         globals: globals, // trusted global variables.
         methodName: methodName
-      }) + '\u0000\u0000'); // the chunk separator
+      });
+      // timestamp object in case of collision
+      if (this.sandcastle.sharedObject.isClosed()) {
+        this.sandcastle.createMmap();
+      }
+      // handle resize issue due to roughly 1/3 of free memory
+      if (((Buffer.byteLength(pack) + 400)*3) > this.sandcastle.sharedObject.get_free_memory()) {
+        this.onExit(methodName, EJSON.stringify({
+          error: { message: 'Script and Payload exceeds your memory limit.' }
+        }));
+        return;
+      }
+      pack = Buffer.from(pack);
+      const nanoSecs = process.hrtime();
+      this.timeStamp = ((nanoSecs[0]*1e9) + nanoSecs[1]).toString(36);
+      this.sandcastle.sharedObject[`type${this.timeStamp}`] = 'script';
+      this.sandcastle.sharedObject[`sandScript${this.timeStamp}`] = pack;
+      client.write(`${this.sandcastle.sharedFileID}-${this.timeStamp}`);
     });
 
-    client.on('close', function(hadError) {
+    client.on('close', (hadError) => {
       if (hadError) { client.destroy(); }
       client.removeAllListeners();
-      stream.end();
-      stream.removeAllListeners();
-      stream.reset();
-      if (!_this.dataReceived) {
-        setTimeout(function() {
-          _this.createClient(methodName, globals);
-        }, 500);
+      if (!this.dataReceived) {
+        setTimeout(() => {
+          this.createClient(methodName, globals);
+        }, 300);
       }
     });
 
-    client.on('error', function(err) {
-      setTimeout(function() {
-        _this.createClient(methodName, globals);
+    client.on('error', () => {
+      setTimeout(() => {
+        this.createClient(methodName, globals);
       }, 500);
     });
 
-    stream.split('\u0000\u0000', function(chunk) {
-      client.end();
-      _this.onExit(methodName, chunk);
-    });
-
-    stream.split('\u0000', function (chunk) {
-      _this.onTask(client, methodName, chunk); // handling the task
-    });
-
-    client.on('data', function(chunk) {
-      _this.dataReceived = true;
-      stream.write(chunk);
+    client.on('data', (data) => {
+      this.dataReceived = true;
+      data = data.toString();
+      const ufid = data.substring(0, data.indexOf('-'));
+      const timeStamp = data.substring(data.indexOf('-')+1);
+      this.readOnlySharedObject = new MMAP.Open(`/tmp/shared-${ufid}`);
+      if (this.readOnlySharedObject[`type${timeStamp}`]  === 'script') {
+        client.end();
+        const sharedData = this.readOnlySharedObject[`sandScript${timeStamp}`].toString();
```

Comment:
why do you need to call `toString()` on this line and line 127, but you don't need to call toString() on lines 30 and 32 of `lib/sandbox.js`?

Reply:
there is a toString on line 24 of sandbox.js. this is due to the write sending as a buffer